lasagne/layers/normalization.py

```diff
@@ -373,3 +375,144 @@ def batch_norm(layer, **kwargs):
+        the square root and dividing by it, to avoid numerical problems
+    **kwargs
+        Any additional keyword arguments are passed to the :class:`Layer`
+        superclass.
+    Notes
+    -----
+    This layer should be inserted between a linear transformation (such as a
+    :class:`DenseLayer`, or :class:`Conv2DLayer`) and its nonlinearity. The
+    convenience function :func:`instance_norm` modifies an existing layer to
+    insert instance normalization in front of its nonlinearity.
+    The behavior can be controlled by passing keyword arguments to
+    :func:`lasagne.layers.get_output()` when building the output expression
+    of any network containing this layer.
+    During training, [1]_ normalize each instance in an input mini-batch
+    by its statistics.
+    See also
+    --------
+    batch_norm : Convenience function to apply batch normalization to a layer
+    References
+    ----------
+    .. [1] Ulyanov, D., Vedaldi, A., & Lempitsky, V. (2016):
+           Instance Normalization: The Missing Ingredient for Fast Stylization.
+           https://arxiv.org/pdf/1607.08022.pdf.
+    """
+    def __init__(self, incoming, axes='auto', epsilon=1e-4, **kwargs):
+        super(InstanceNormLayer, self).__init__(incoming, **kwargs)
+
+        if axes == 'auto':
+            # default: normalize over spatial dimensions only,
+            # i.e. separate for each instance in the batch
+            axes = (2, 3)
+        elif isinstance(axes, int):
+            axes = (axes,)
+        self.axes = axes
+
+        self.epsilon = epsilon
+
+    def get_output_for(self, input, **kwargs):
+        input_mean = input.mean(self.axes)
+        input_inv_std = T.inv(T.sqrt(input.var(self.axes) + self.epsilon))
+
+        mean = input_mean
+        inv_std = input_inv_std
+
+        # prepare dimshuffle pattern inserting broadcastable axes as needed
+        param_axes = iter(range(input.ndim - len(self.axes)))
+        pattern = ['x' if input_axis in self.axes
+                   else next(param_axes)
+                   for input_axis in range(input.ndim)]
+
+        # apply dimshuffle pattern to all parameters
+        # beta = 0 if self.beta is None else self.beta.dimshuffle(pattern)
+        # gamma = 1 if self.gamma is None else self.gamma.dimshuffle(pattern)
+        mean = mean.dimshuffle(pattern)
+        inv_std = inv_std.dimshuffle(pattern)
+
+        normalized = (input - mean) * inv_std
+
+        return normalized
+
+
+def instance_norm(layer, **kwargs):
+    """
+    Apply instance normalization to an existing layer. This is a convenience
+    function modifying an existing layer to include instance normalization: It
+    will steal the layer's nonlinearity if there is one (effectively
+    introducing the normalization right before the nonlinearity), remove
+    the layer's bias if there is one (because it would be redundant), and add
+    a :class:`InstanceNormLayer` and :class:`NonlinearityLayer` on top.
```

Comment:
> The instance normalization paper does not mention the bias whatsoever, suggesting that they don't remove it and plainly apply instance norm on top of the layers output. 

But that would subtract the bias again. I doubt that's a good idea. Did they provide any code?

> I added beta and gamma as learnable parameters, which are broadcast across the batch-size as well as the trailing dimensions beyond the 2nd dimension.

This fits the layer normalization paper, but it's a bit sad we can't control the axes then. I'm torn between leaving it as it is, or not providing beta/gamma at all and offloading this to a ScaleLayer and BiasLayer on top (which could be added automatically by the convenience function if asked for). This would simplify the layer and tests. For the batch normalization layer it was good to have everything in a single layer, because this allows to use cuDNN's batch normalization primitives. For instance normalization, I doubt this will play a role.